domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresConnectionManager.java

```diff
@@ -0,0 +1,128 @@
+    this.tryRemoveOldestAvailableConnectionAndClose();
+  }
+
+  public Future<PgConnection> getConnection(Pool pool, String schemaName, String tenantId) {
+    if (! PostgresClient.sharedPgPool) {
+      LOG.debug("Not in shared pool mode");
+      return pool.getConnection().map(PgConnection.class::cast);
+    }
+
+    LOG.debug("In shared pool mode");
+    return getOrCreateCachedConnection(pool, schemaName, tenantId);
+  }
+
+  private void tryAddToCache(CachedPgConnection connection) {
+    if (connectionCache.contains(connection)) {
+      LOG.debug("Item already exists in cache: {} {} {}",
+          connection.getTenantId(), connection.getSessionId(), connection.isAvailable());
+      return;
+    }
+    connectionCache.add(connection);
+  }
+
+  private void tryRemoveOldestAvailableConnectionAndClose() {
+    // The cache must not grow larger than the max pool size of the underlying pool.
+    // TODO get the pool size from the env.
+    var cacheExhausted = this.connectionCache.size() >= PostgresClient.DEFAULT_MAX_POOL_SIZE;
+    if (! cacheExhausted) {
+      LOG.debug("Cache is not yet exhausted");
+      return;
+    }
+
+    connectionCache.stream()
+        .filter(CachedPgConnection::isAvailable)
+        .min(Comparator.comparingLong(CachedPgConnection::getTimestamp))
+        .ifPresent(connection -> {
+          LOG.debug("Removing and closing oldest available connection: {} {}",
+              connection.getTenantId(), connection.getSessionId());
+          if (connection.getWrappedConnection() != null) {
+            activeConnectionCount--;
+            connection.getWrappedConnection().close();
+          }
+          connectionCache.remove(connection);
+        });
+  }
+
+  private void logCache(String context, int poolSize) {
+    LOG.debug("Current cache state: {}", context);
+
+    synchronized (connectionCache) {
+      connectionCache.forEach(c ->
+        LOG.debug("{} {} {} {} {} {} {} {} {}", cacheHits, cacheMisses, connectionCache.size(),
+            activeConnectionCount, poolSize, c.getSessionId(), c.getTenantId(), c.isAvailable(),
+            c.getConnectionHash()));
+    }
+  }
+
+  private Future<PgConnection> getOrCreateCachedConnection(Pool pool, String schemaName, String tenantId) {
+    Optional<CachedPgConnection> connectionOptional =
+        connectionCache.stream().filter(item ->
+            item.getTenantId().equals(tenantId) && item.isAvailable()).findFirst();
+    if (connectionOptional.isPresent()) {
+      cacheHits++;
+      CachedPgConnection connection = connectionOptional.get();
+      var ctx = String.format("cache hit %s %s", connection.getTenantId(), connection.getSessionId());
+      logCache(ctx, pool.size());
+      connection.setUnavailable();
+      return Future.succeededFuture(connection);
+    }
+
+    var ctx = String.format("cache miss %s", tenantId);
+    cacheMisses++;
+    logCache(ctx, pool.size());
+
+    return createConnectionSession(pool, schemaName, tenantId);
```

Comment:
My preference is not to implement our own wait queue since the current code has been tested and is known to work, adding this additional feature will add time (another round of testing) and risk (potential unforeseen problems) to this implementation. Perhaps we can tackle this in a phase 2 if performance of the shared pool is thought to need further enhancements.